### PR TITLE
Issue #7687:  Adding code examples for ImportOrder check

### PR DESCRIPTION
**Pull request by aryaniiit002**
Issue #7687:  Adding code examples for ImportOrder check

### CODE REVIEW

**Comment by aryaniiit002** [comment]
[INFO] Scanned 2169 class file(s) for forbidden API invocations (in 16.05s), 0 error(s).
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  17:57 min
[INFO] Finished at: 2021-01-01T08:07:03+05:30
[INFO] ------------------------------------------------------------------------

**Comment by strkkk** [comment]
@aryaniiit002 please fix CI failures

**Comment by aryaniiit002** [comment]
> @aryaniiit002 please fix CI failures

yes sir

**Comment by aryaniiit002** [comment]
i could use a little help here.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 1]
```diff
@@ -140,6 +140,32 @@
...
+ * import javax.net.ssl.TrustManager; //OK
+ * import javax.net.ssl.X509TrustManager; //OK
+ *
+ * import org.apache.http.conn.ClientConnectionManager; //OK
+ * import org.apache.http.conn.scheme.Scheme; //OK
+ *
+ * import com.neurologic.http.HttpClient; //OK
+ * import com.neurologic.http.impl.ApacheHttpClient; //OK
+ *
+ * public class SomeClass { ... }
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:166]: It is a good idea use compilable code in examples.  A good workflow that I use when updating check documentation is very similar to creating bug reports:

1. Create the configuration
2. Create the the file to be checked
3. Make sure code compiles
4. Run checkstyle on the file
5. place "// ok" and "// violation" comments where appropriate
6. Paste into documentation

This will help you to learn Checkstyle and develop a good work flow.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 2]
```diff
@@ -148,7 +174,8 @@
  * group of static imports is on the top
  * </li>
  * <li>
- * groups of type imports: "java" and "javax" packages first, then "org" and then all other imports
+ * groups of type imports: "java" and "javax" packages first,
+ * then "org" and then all other imports
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:178]: Please revert this change.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 3]
```diff
@@ -140,6 +140,32 @@
...
+ * import java.net.URL; //OK
+ * import java.security.KeyManagementException; //OK
+ *
+ * import javax.net.ssl.SSLContext; //OK
+ * import javax.net.ssl.TrustManager; //OK
+ * import javax.net.ssl.X509TrustManager; //OK
+ *
+ * import org.apache.http.conn.ClientConnectionManager; //OK
+ * import org.apache.http.conn.scheme.Scheme; //OK
+ *
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:162]: Please show some violations here, also this example probably doesn't need to have this many packages.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 4]
```diff
@@ -140,6 +140,32 @@
  * </li>
  * </ul>
  * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="ImportOrder"/&gt;
+ * </pre>
+ * <p>
+ * Example of Import order:
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:149]: Just "Example:" is good here.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 5]
```diff
@@ -158,16 +185,36 @@
...
+ * import javax.net.ssl.TrustManager; //OK
+ *
+ * import javax.net.ssl.X509TrustManager; //Violation; Groups should not separate internally
+ *
+ * import org.apache.http.conn.ClientConnectionManager; //OK
+ * import org.apache.http.conn.scheme.Scheme; //OK
+ *
+ * public class SomeClass { ... }
+ * </pre>
+ * <p>
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:207]: This example should follow the configuration on lines 221-226.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 6]
```diff
@@ -140,6 +140,32 @@
...
+ * Example of Import order:
+ * </p>
+ * <pre>
+ * import java.io.IOException; //OK
+ * import java.net.URL; //OK
+ * import java.security.KeyManagementException; //OK
+ *
+ * import javax.net.ssl.SSLContext; //OK
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:156]: Make sure to leave a space after comment delimiter:
like this: `// OK`, not `//OK`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 7]
```diff
@@ -158,16 +185,36 @@
...
+ * </pre>
+ * <p>
  * Notes:
  * </p>
  * <ul>
  * <li>
- * "com" package is not mentioned on configuration, because it is ignored by Eclipse Kepler and Luna
- * (looks like Eclipse defect)
+ * "com" package is not mentioned on configuration, because it is
+ * ignored by Eclipse Kepler and Luna (looks like Eclipse defect)
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:213]: Revert this change

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 8]
```diff
@@ -158,16 +185,36 @@
...
  * </li>
  * <li>
- * configuration below doesn't work in all 100% cases due to inconsistent behavior prior to
- * Mars release, but covers most scenarios
+ * configuration below doesn't work in all 100% cases due to inconsistent behavior
+ * prior to Mars release, but covers most scenarios
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:217]: Revert

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 9]
```diff
@@ -180,16 +227,16 @@
  * &lt;/module&gt;
  * </pre>
  * <p>
- * To configure the check so that it matches default Eclipse formatter configuration
- * (tested on Mars release):
+ * To configure the check so that it matches default Eclipse formatter
+ * configuration (tested on Mars release):
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:231]: Revert

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 10]
```diff
@@ -180,16 +227,16 @@
...
  * </p>
  * <ul>
  * <li>
  * group of static imports is on the top
  * </li>
  * <li>
- * groups of type imports: "java" and "javax" packages first, then "org" and "com",
- * then all other imports as one group
+ * groups of type imports: "java" and "javax" packages first,
+ * then "org" and "com", then all other imports as one group
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:239]: Revert

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 11]
```diff
@@ -208,15 +255,43 @@
...
+ * configuration (tested on v2018.2):
  * </p>
  * <ul>
  * <li>
  * group of static imports is on the bottom
  * </li>
  * <li>
- * groups of type imports: all imports except of "javax" and "java", then "javax" and "java"
+ * groups of type imports: all imports except of "javax"
+ * and "java", then "javax" and "java"
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:294]: Revert

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 12]
```diff
@@ -229,13 +304,13 @@
...
- * other groups.
+ * ImportOrder has a limitation by design to enforce an empty line between groups
+ * ("java", "javax"). There is no flexibility to enforce empty lines between some
+ * groups and no empty lines between other groups.
  * </p>
  * <p>
- * Note: "separated" option is disabled because IDEA default has blank line between "java" and
- * static imports, and no blank line between "javax" and "java".
+ * Note: "separated" option is disabled because IDEA default has blank line
+ * between "java" and static imports, and no blank line between "javax" and "java".
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:313]: Revert

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 13]
```diff
@@ -251,6 +326,22 @@
...
+ *
+ * import java.net.URL; //OK
+ * import java.security.KeyManagementException; //OK
+ * import javax.net.ssl.TrustManager; //OK; no blank line between "javax" and "java".
+ *
+ * import static java.awt.Button.A; //Violation; Should follow 'sortStaticImportsAlphabetically'
+ * import static java.lang.Math.PI; //OK
+ * import static java.lang.Math.abs; //OK, alphabetical case sensitive ASCII order, 'P' &lt; 'a'
+ *
+ * public class SomeClass { ... }
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:342]: Code should compile

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 14]
```diff
@@ -158,16 +185,36 @@
...
+ * import java.security.KeyManagementException; //OK
+ *
+ * import javax.net.ssl.TrustManager; //OK
+ *
+ * import javax.net.ssl.X509TrustManager; //Violation; Groups should not separate internally
+ *
+ * import org.apache.http.conn.ClientConnectionManager; //OK
+ * import org.apache.http.conn.scheme.Scheme; //OK
+ *
+ * public class SomeClass { ... }
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:205]: Code should compile

**Review by nmancus1 — CHANGES_REQUESTED** [review]
@aryaniiit002 thanks for your contribution! Note that I have only reviewed `ImportOrderCheck.java`, but all comments apply to `config_imports.xml` since they have to match (as you have already found out). Some items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 13 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:342]: could you please elaborate this?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 4 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:149]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 6 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:156]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 3 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:162]: ok no violations here bcoz this piece is for default importOrder values.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 1 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:166]: how do i check a file?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 2 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:178]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 7 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:213]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 8 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:217]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 9 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:231]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 10 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:239]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 5 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:207]: sorry i didnt get this.
are static group and java group should not be separated?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 11 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:294]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 12 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:313]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 1 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:183]: > how do i check a file?
https://checkstyle.sourceforge.io/cmdline.html

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 12 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:368]: I revert again bcoz of this ERROR:  ....... ......  "Line should not be longer than 100 symbols [lineLength] "

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:368]: @aryaniiit002 from newest master:
```
➜  checkstyle git:(master) awk '{print length, $0}'  src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java |sort -nr|head -2 
110  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#STATIC_IMPORT">
100      * Specify list of <b>static</b> import groups (every group identified either by a common prefix

```
The only existing line in this file that is over 100 characters is line 138, and Checkstyle makes exceptions for links. Somehow extra characters must be getting added if this is failing. If you ever have any doubts about your changes, you can use [git diff](https://git-scm.com/docs/git-diff) to see the difference between your branch's code, and master.

**Comment by aryaniiit002** [comment]
something weird happened all the changes i have made in the code after your comments are mixed with the code  that was before your comments as you can see .

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 15]
```diff
@@ -140,6 +140,49 @@
  * </li>
  * </ul>
  * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name="ImportOrder"/&gt;
+ * </pre>
+ * <p>
+<<<<<<< HEAD
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:149]: This is left from git conflict resolution, please remove.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java [hunk 16]
```diff
@@ -140,6 +140,49 @@
...
+ *
+ * import javax.net.ssl.SSLContext; // OK
+ * import javax.net.ssl.TrustManager; // OK
+ *
+ * import org.apache.http.conn.ClientConnectionManager; // OK
+ * import org.apache.http.conn.scheme.Scheme; // OK
+ *
+ * import com.neurologic.http.HttpClient; // OK
+ * import com.neurologic.http.impl.ApacheHttpClient; // OK
+=======
```
Comment (nmancus1) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:164]: This is left from git conflict resolution, please remove.

**Comment by nmancus1** [comment]
> something weird happened all the changes i have made in the code after your comments are mixed with the code as it was before your comments as you can see .

I think you dropped some of your changes after rebasing. A good workflow that I use is this:
1. `git checkout` this branch
2. `git reset HEAD^` this will undo the last commit, but keep your changes.  If you are using an IDE, you will see the lines where your changes have been made, and can just use the rollback function to remove changes that you have been asked to revert:
image

3.  Make commit again.
4. rebase on newest master
4. `mvn clean verify`, make sure it passes
5. `git push --force`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 15 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:149]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 12 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:368]: OKay

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 15 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:149]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 12 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:368]: idk how but i am still getting this error.

**Comment by aryaniiit002** [comment]
> > something weird happened all the changes i have made in the code after your comments are mixed with the code as it was before your comments as you can see .
> 
> I think you dropped some of your changes after rebasing. A good workflow that I use after changes are requested is this:
> 
>     1. `git checkout` this branch
> 
>     2. `git reset HEAD^` this will undo the last commit, but keep your changes.  If you are using an IDE, you will see the lines where your changes have been made, and can just use the rollback function to remove changes that you have been asked to revert:
>        image
> 
>     3. Make other changes.
> 
>     4. Make commit again.
> 
>     5. Rebase on newest master
> 
>     6. `mvn clean verify`, make sure it passes
> 
>     7. `git push --force`

i m sorry for the late reply my laptop cannot handle browser and intelliJ same time

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 1 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:183]: that javac code

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java (hunk 14 above)
Comment (aryaniiit002) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/ImportOrderCheck.java:237]: pls tell me how do check this?
check whether code is compiling or not? You said something about javac my code in a file i am not sure how to do that.

**Comment by aryaniiit002** [comment]
this is an experiment.

**Comment by aryaniiit002** [comment]
Diff Regression config: https://gist.githubusercontent.com/aryaniiit002/86e8279961bac2a91c0b6b3d948d6e60/raw/179ea72fb4b227d0954fd194f540f22e7eb3ca25/checkstyle_checks.xml
Diff Regression projects: https://gist.githubusercontent.com/aryaniiit002/97577199a93846562de23a0d6304e496/raw/1e822753136da53287c702a879090c63061936f2/projects-to-test-on.properties

**Comment by aryaniiit002** [comment]
GitHub, generate report

**Comment by aryaniiit002** [comment]
Github, generate website
